dpe/src/x509.rs

```diff
@@ -483,6 +563,136 @@ impl X509CertWriter<'_> {
+    ///
+    /// https://trustedcomputinggroup.org/wp-content/uploads/TCG_DICE_Attestation_Architecture_r22_02dec2020.pdf
+    ///
+    /// TcbInfo makes use of implicitly encoded types. This means the tag
+    /// denotes that the type is implicit (8th bit set) and number of the
+    /// field. For example, "Implicit tag number 2" would be encoded with
+    /// the tag 0x82 for primitive types.
+    ///
+    /// For constructed types (SEQUENCE, SEQUENCE OF, SET, SET OF) the 6th
+    /// bit is also set. For example, "Implicit tag number 2" would be encoded
+    /// with tag 0xA2 for constructed types.
+    fn encode_tcb_info(&mut self, node: &TciNodeData) -> Result<usize, DpeErrorCode> {
+        let tcb_info_size = Self::get_tcb_info_size(node, /*tagged=*/ false)?;
+        // TcbInfo sequence
+        let mut bytes_written = self.encode_byte(Self::SEQUENCE_TAG)?;
+        bytes_written += self.encode_size_field(tcb_info_size)?;
+
+        // fwids SEQUENCE OF
+        // IMPLICIT [6] Constructed
+        let fwid_size = Self::get_fwid_size(&node.tci_current.0, /*tagged=*/ true)?;
+        bytes_written += self.encode_byte(0xA6)?;
+        bytes_written += self.encode_size_field(fwid_size * 2)?;
+
+        // fwid[0] current measurement
+        bytes_written += self.encode_fwid(&node.tci_current)?;
+
+        // fwid[1] journey measurement
+        bytes_written += self.encode_fwid(&node.tci_cumulative)?;
+
+        // vendorInfo OCTET STRING
+        // IMPLICIT[8] Primitive
+        let vinfo = if node.flag_is_internal() {
+            "VNDR"
+        } else {
+            "USER"
+        };
+        bytes_written += self.encode_byte(0x88)?;
+        bytes_written += self.encode_size_field(vinfo.len())?;
+        bytes_written += self.encode_bytes(vinfo.as_bytes())?;
+
+        // type OCTET STRING
+        // IMPLICIT[9] Primitive
+        bytes_written += self.encode_byte(0x89)?;
+        bytes_written += self.encode_size_field(core::mem::size_of::<u32>())?;
+        bytes_written += self.encode_bytes(&node.tci_type.to_be_bytes())?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a tcg-dice-MultiTcbInfo extension
+    ///
+    /// https://trustedcomputinggroup.org/wp-content/uploads/TCG_DICE_Attestation_Architecture_r22_02dec2020.pdf
+    fn encode_multi_tcb_info(
+        &mut self,
+        measurements: &MeasurementData,
+    ) -> Result<usize, DpeErrorCode> {
+        let tcb_info_size =
+            Self::get_tcb_info_size(measurements.tci_nodes[0], /*tagged=*/ true)?;
+        let tcb_infos_size = tcb_info_size * measurements.tci_nodes.len();
+        let multi_tcb_info_size = Self::get_structure_size(tcb_infos_size, /*tagged=*/ true)?;
+        let extension_size = Self::get_structure_size(Self::MULTI_TCBINFO_OID.len(), /*tagged=*/true)? +
+            Self::get_structure_size(1, /*tagged=*/true)? + // bool size
+            Self::get_structure_size(multi_tcb_info_size, /*tagged=*/true)?; // Octet string
```

Comment:
Can you use the helper functions for these?
```suggestion
        let multi_tcb_info_size = Self::get_multi_tcb_info_size(measurements, /*tagged=*/ true)?;
        let extension_size = Self::get_extensions_size(measurements, /*tagged=*/true)?; // Octet string
```

Reply:
Will do. I did it this way to avoid recomputing the same intermediate values, but it's probably better to do this for clarity instead of saving a few cycles.

Reply:
SG. I think it should all be inlined anyway, so it shouldn't change performance.